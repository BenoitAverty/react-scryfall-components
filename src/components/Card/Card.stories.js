import React from 'react';

import { storiesOf } from '@storybook/react';

import Card from '.';

storiesOf('Card', module)
  .add('Card', () => <Card id="eb28b35c-28a5-4042-b21d-6d43658a16eb" />)
  .add('different sizes', () => (
    <div>
      <Card id="eb28b35c-28a5-4042-b21d-6d43658a16eb" size={Card.SIZE_SMALL} />
      <Card id="eb28b35c-28a5-4042-b21d-6d43658a16eb" size={Card.SIZE_NORMAL} />
      <Card id="eb28b35c-28a5-4042-b21d-6d43658a16eb" size={Card.SIZE_LARGE} />
    </div>
  ));
