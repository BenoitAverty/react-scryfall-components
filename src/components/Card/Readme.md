Basic card:

```js
<Card id="eb28b35c-28a5-4042-b21d-6d43658a16eb" />
```

Different sizes:
```js
<Card id="eb28b35c-28a5-4042-b21d-6d43658a16eb" size={Card.SIZE_SMALL} />
<Card id="eb28b35c-28a5-4042-b21d-6d43658a16eb" size={Card.SIZE_NORMAL} />
<Card id="eb28b35c-28a5-4042-b21d-6d43658a16eb" size={Card.SIZE_LARGE} />
```