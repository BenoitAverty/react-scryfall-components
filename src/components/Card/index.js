import React from 'react';
import PropTypes from 'prop-types';
import axios from '../../utils/axios';

/**
 * Renders a card image from scryfall.
 */
class Card extends React.Component {
  static displayName = 'Card';

  static SIZE_SMALL = 'small';
  static SIZE_NORMAL = 'normal';
  static SIZE_LARGE = 'large';

  static propTypes = {
    /**
     * Scryfall ID of the card to render. This is a uuid that should look like this : eb28b35c-28a5-4042-b21d-6d43658a16eb
     */
    id: PropTypes.string.isRequired,
    /**
     * Size of the image to render
     */
    size: PropTypes.oneOf([Card.SIZE_SMALL, Card.SIZE_NORMAL, Card.SIZE_LARGE]),
  };

  static defaultProps = {
    size: Card.SIZE_NORMAL,
  };

  constructor(props) {
    super(props);

    this.state = {
      imgUrl: null,
    };
  }

  componentDidMount() {
    const { id, size } = this.props;

    axios.get(`/cards/${id}`).then(resp => {
      this.setState({
        imgUrl: resp.data.image_uris[size],
        cardName: resp.data.name,
      });
    });
  }

  render() {
    const { imgUrl, cardName } = this.state;

    return imgUrl ? <img src={imgUrl} alt={cardName} /> : null;
  }
}

export default Card;
