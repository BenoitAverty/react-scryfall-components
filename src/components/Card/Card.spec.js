// Import all that is needed for testing
import React from 'react';
import { render, cleanup, waitForElement } from 'react-testing-library';
import 'jest-dom/extend-expect';

// Import the component under test
import Card from '.';

describe('Card component', () => {
  beforeEach(() => {
    cleanup();
  });

  describe('Rendering a card image', () => {
    test('snapshot', async () => {
      const { container } = render(<Card id="eb28b35c-28a5-4042-b21d-6d43658a16eb" />);

      // Wait for an img to appear
      await waitForElement(() => document.querySelector('img'), { container });

      expect(container.firstChild).toMatchSnapshot();
    });

    it('renders an image if given a scryfall id', async () => {
      // Crucible of worlds
      const { container } = render(<Card id="eb28b35c-28a5-4042-b21d-6d43658a16eb" />);

      // Wait for an img to appear
      await waitForElement(() => container.querySelector('img'), { container });

      expect(container.querySelectorAll('img')).toHaveLength(1);
      expect(container.querySelector('img')).toHaveAttribute('src', 'https://img.scryfall.com/cards/normal/en/m19/229.jpg?1531451273');
      expect(container.querySelector('img')).toHaveAttribute('alt', 'Crucible of Worlds');
    });

    it('renders the small image with the size property at small', async () => {
      // Momentary blink
      const { container } = render(<Card id="4206d5f8-edeb-4e0b-8e98-736f6ccdcf99" size={Card.SIZE_SMALL} />);

      // Wait for an img to appear
      await waitForElement(() => container.querySelector('img'), { container });

      expect(container.querySelectorAll('img')).toHaveLength(1);
      expect(container.querySelector('img')).toHaveAttribute('src', 'https://img.scryfall.com/cards/small/en/mm3/16.jpg?1517813031');
      expect(container.querySelector('img')).toHaveAttribute('alt', 'Momentary Blink');
    });
  });
});
