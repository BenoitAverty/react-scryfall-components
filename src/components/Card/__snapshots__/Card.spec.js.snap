// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component Rendering a card image snapshot 1`] = `
<img
  alt="Crucible of Worlds"
  src="https://img.scryfall.com/cards/normal/en/m19/229.jpg?1531451273"
/>
`;
